tezos/new_context/src/gc/worker.rs

```diff
@@ -0,0 +1,183 @@
+    pub(crate) cycles: Cycles,
+    pub(crate) free_ids: Producer<HashId>,
+    pub(crate) recv: Receiver<Command>,
+    pub(crate) pending: Vec<HashId>,
+}
+
+pub(crate) enum Command {
+    StartNewCycle {
+        values_in_cycle: BTreeMap<HashId, Option<Arc<[u8]>>>,
+        new_ids: Vec<HashId>,
+    },
+    MarkReused {
+        reused: Vec<HashId>,
+    },
+}
+
+pub(crate) struct Cycles {
+    list: VecDeque<BTreeMap<HashId, Option<Arc<[u8]>>>>,
+}
+
+impl Default for Cycles {
+    fn default() -> Self {
+        let mut list = VecDeque::with_capacity(PRESERVE_CYCLE_COUNT);
+
+        for _ in 0..PRESERVE_CYCLE_COUNT {
+            list.push_back(Default::default());
+        }
+
+        Self { list }
+    }
+}
+
+impl Cycles {
+    fn move_to_last_cycle(&mut self, hash_id: HashId) -> Option<Arc<[u8]>> {
+        let len = self.list.len();
+        let mut value = None;
+
+        for store in &mut self.list.iter_mut().take(len - 1) {
+            if let Some(item) = store.remove(&hash_id).flatten() {
+                value = Some(item);
+            };
+        }
+
+        let value = value?;
+        if let Some(v) = self.list.back_mut() {
+            v.insert(hash_id, Some(Arc::clone(&value)));
+        } else {
+            eprintln!("GC: Failed to insert value in Cycles")
+        }
+
+        return Some(value);
+    }
+
+    fn roll(&mut self, new_cycle: BTreeMap<HashId, Option<Arc<[u8]>>>) -> Vec<HashId> {
+        let unused = self.list.pop_front().unwrap_or_else(BTreeMap::new);
+        self.list.push_back(new_cycle);
+
+        let mut vec = Vec::with_capacity(unused.len());
+        for id in unused.keys() {
+            vec.push(*id);
+        }
+        vec
+    }
+}
+
+impl GCThread {
+    pub(crate) fn run(mut self) {
+        while let Ok(msg) = self.recv.recv() {
+            match msg {
+                Command::StartNewCycle {
+                    values_in_cycle,
+                    new_ids,
+                } => self.start_new_cycle(values_in_cycle, new_ids),
+                Command::MarkReused { reused } => self.mark_reused(reused),
+            }
+        }
+    }
+
+    fn start_new_cycle(
+        &mut self,
+        mut new_cycle: BTreeMap<HashId, Option<Arc<[u8]>>>,
+        new_ids: Vec<HashId>,
+    ) {
+        for hash_id in new_ids.into_iter() {
+            new_cycle.entry(hash_id).or_insert(None);
+        }
+        let unused = self.cycles.roll(new_cycle);
+        self.send_unused(unused);
+    }
+
+    /// Notify the main thread that the ids are free to reused
+    fn send_unused(&mut self, unused: Vec<HashId>) {
+        let unused_length = unused.len();
+        let navailable = self.free_ids.available();
+
+        let (to_send, pending) = if navailable < unused_length {
+            unused.split_at(navailable)
+        } else {
+            (&unused[..], &[][..])
+        };
+
+        if self.free_ids.push_slice(&to_send).is_err() {
+            eprintln!("GC: Fail to send free ids");
```

Comment:
Not sure what the possible errors are here, but `if let Err(err) = ...` to include the error in the output would be helpful.